### PR TITLE
OCPBUGS-18865: add monitortest: in-cluster disruption monitors

### PR DESCRIPTION
**Pull request by vrutkovs**
This re-adds kubeapiserver disruption monitors. Previously we had "is kubeapiserver available externally", but it was insufficient to find out what happens, as components use various ways to reach it. 

The PR adds monitors to collect kube-apiserver/openshift-apiserver disruption data using:
* api-int endpoint on the loadbalancer
* service network
* localhost
by running several pods and collecting data. Later the test will fetch disruption data from the pods, combine it and display it as intervals alongside external disruption query.

Revert of https://github.com/openshift/origin/pull/28274 and reworked to use monitortest framework

### CODE REVIEW

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

File: pkg/monitortests/kubeapiserver/incluster_disruption_serializer/manifests/namespace.yaml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 kind: Namespace
 apiVersion: v1
 metadata:
-  generateName: e2e-disruption-monitor-
+  name: e2e-disruption-monitor
```
Comment (deads2k) [inline pkg/monitortests/kubeapiserver/incluster_disruption_serializer/manifests/namespace.yaml:4]: This is a fix in the wrong direction.  YOu can either add RBAC for the SCC or set the runlevel=0 to skip the SCC check in this namespace.

Comment (vrutkovs) [reply pkg/monitortests/kubeapiserver/incluster_disruption_serializer/manifests/namespace.yaml:4]: I did add RBAC for SCC, but for some reason we run disruption tests twice (once during uprgade and once during conformance tests). Its unreasonable to create a new namespace for the second run

Comment (deads2k) [reply pkg/monitortests/kubeapiserver/incluster_disruption_serializer/manifests/namespace.yaml:4]: > Its unreasonable to create a new namespace and new RBAC rules for the second run

On what basis?

New namespaces prevent overlap and allow multiple watchers.  Collapsing to a single namespace is an anti-goal.  I'm ok with RBAC if that's what you prefer.  I'm also ok with runlevel=0.

Comment (vrutkovs) [reply pkg/monitortests/kubeapiserver/incluster_disruption_serializer/manifests/namespace.yaml:4]: There is no reason to run multiple disruption watchers for the same apiservers in parallel

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-job periodic-ci-openshift-release-master-nightly-4.14-e2e-metal-ipi-ovn-ipv6

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-azure-sdn-upgrade 5

**Comment by smg247** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-azure-sdn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/retest

Ready for review. `runlevel=0` helped prevent SCC errors, namespace handling updated to make sure its being cleaned up properly (also uses `generateName` too)

**Comment by vrutkovs** [comment]
/hold

invalid permissions for serviceaccount - ClusterRole doesn't seem to have applied

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade

**Comment by vrutkovs** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade

**Comment by vrutkovs** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade

**Comment by vrutkovs** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-aggregate periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade 5

**Comment by vrutkovs** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade

**Comment by vrutkovs** [comment]
/retest
/hold cancel

All issues seems fixed (it was colliding with resource watcher), ready for review

**Comment by vrutkovs** [comment]
/payload-job periodic-ci-openshift-release-master-ci-4.15-upgrade-from-stable-4.14-e2e-aws-ovn-upgrade

**Comment by vrutkovs** [comment]
/payload-job periodic-ci-openshift-release-master-nightly-4.15-e2e-metal-ipi-ovn-ipv6

File: pkg/monitortestlibrary/disruptionlibrary/collect.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+	"github.com/openshift/origin/pkg/monitor/monitorapi"
+	monitorserialization "github.com/openshift/origin/pkg/monitor/serialization"
+	"github.com/openshift/origin/pkg/test/ginkgo/junitapi"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+)
+
+func CollectIntervalsForPods(ctx context.Context, kubeClient kubernetes.Interface, namespace string, labelSelector labels.Selector) (monitorapi.Intervals, []*junitapi.JUnitTestCase, []error) {
```
[the remark this answers is not in the document]
Comment (vrutkovs) [reply pkg/monitortestlibrary/disruptionlibrary/collect.go:19]: Good idea, updated pna to reuse that function

File: pkg/defaultmonitortests/types.go [hunk 3]
```diff
@@ -119,6 +121,7 @@ func newDefaultMonitorTests(info monitortestframework.MonitorTestInitializationI
 	monitorTestRegistry.AddMonitorTestOrDie("external-azure-cloud-service-availability", "Test Framework", disruptionexternalazurecloudservicemonitoring.NewCloudAvailabilityInvariant())
 	monitorTestRegistry.AddMonitorTestOrDie("pathological-event-analyzer", "Test Framework", pathologicaleventanalyzer.NewAnalyzer())
 	monitorTestRegistry.AddMonitorTestOrDie("disruption-summary-serializer", "Test Framework", disruptionserializer.NewDisruptionSummarySerializer())
+	monitorTestRegistry.AddMonitorTestOrDie("incluster-disruption-serializer", "kube-apiserver", incluster_disruption_serializer.NewInvariantInClusterDisruption(info))
```
[the remark this answers is not in the document]
Comment (vrutkovs) [reply pkg/defaultmonitortests/types.go]: updated

File: pkg/monitortestframework/impl.go [hunk 4]
```diff
@@ -446,3 +452,36 @@ func (r *monitorTestRegistry) AddRegistryOrDie(registry MonitorTestRegistry) {
 func (r *monitorTestRegistry) getMonitorTests() map[string]*monitorTesttItem {
 	return r.monitorTests
 }
+
+// GetOpenshiftTestsImagePullSpec returns the pull spec, a not-supported-reason, or an error.
+func (m *MonitorTestInitializationInfo) GetOpenshiftTestsImagePullSpec(ctx context.Context, adminRESTConfig *rest.Config) (string, string, error) {
```
[the remark this answers is not in the document]
Comment (vrutkovs) [reply pkg/monitortestframework/impl.go]: No longer required, removed

File: pkg/monitortests/kubeapiserver/incluster_disruption_serializer/remote.go [hunk 5]
```diff
@@ -0,0 +1,421 @@
+package incluster_disruption_serializer
```
[the remark this answers is not in the document]
Comment (vrutkovs) [reply pkg/monitortests/kubeapiserver/incluster_disruption_serializer/remote.go]: >This package and file should probably be renamed. 

Renamed to match other 

>This isn't really a serializer, it doesn't even write artifacts content (handled by the generic disruption serializer), and convention is for these to be in a monitortest.go.

It does, it creates "AdditionalEvents__" files, but I don't mind it to be renamed

File: pkg/monitortests/kubeapiserver/incluster_disruption_serializer/remote.go [hunk 6]
```diff
@@ -0,0 +1,421 @@
...
+}
+
+func (i *InvariantInClusterDisruption) createDeploymentAndWaitToRollout(ctx context.Context, deploymentObj *appsv1.Deployment) error {
+
+	client := i.kubeClient.AppsV1().Deployments(deploymentObj.Namespace)
+	_, err := client.Create(ctx, deploymentObj, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("error creating deployment %s: %v", deploymentObj.Namespace, err)
+	}
+	fmt.Printf("in-cluster monitor: deployment %s:\n%#v\n", deploymentObj.Name, deploymentObj.ObjectMeta)
```
[the remark this answers is not in the document]
Comment (vrutkovs) [reply pkg/monitortests/kubeapiserver/incluster_disruption_serializer/remote.go]: Done

**Pull request merged**